File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        if (strs.empty()) {
            return {};
        }

        unordered_map<string, vector<string>> countMap;

        for (const string& s : strs) {
            vector<int> freq(26, 0);
            for (char c : s) {
                freq[c - 'a']++;
            }

            string key = "";
            char label = 'a';
            for (int count : freq) {
                key += label;
                key += to_string(count);
                label++;
            }

            countMap[key].push_back(s);
        }

        vector<vector<string>> result;
        for (auto& pair : countMap) {
            result.push_back(move(pair.second));
        }

        return result;
    }
};

File: 0049-group-anagrams/0049-group-anagrams.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        if not strs:
            return []
        
        anagram_map = defaultdict(list)
        
        for s in strs:
            freq = [0] * 26
            for char in s:
                freq[ord(char) - ord('a')] += 1
            
            key = tuple(freq)
            anagram_map[key].append(s)
            
        return list(anagram_map.values())